### PR TITLE
Attestation Using Head State instead of Latest State

### PR DESCRIPTION
**Pull request by terenc3t**
This PR fixes the incorrect `AttestationDataResponse` from attester server to validator. Attester server was using **current state (last processed state**) to construct these attestation but it is suppose to use **head state**.

### CODE REVIEW

File: beacon-chain/rpc/attester_server.go [hunk 1]
```diff
@@ -46,22 +45,15 @@ func (as *AttesterServer) AttestationDataAtSlot(ctx context.Context, req *pb.Att
 	if err != nil {
 		return nil, fmt.Errorf("failed to retrieve chain head: %v", err)
 	}
-	blockRoot, err := hashutil.HashBeaconBlock(head)
+	headRoot, err := hashutil.HashBeaconBlock(head)
 	if err != nil {
 		return nil, fmt.Errorf("could not tree hash beacon block: %v", err)
 	}
-	beaconState, err := as.beaconDB.State(ctx)
+	headState, err := as.beaconDB.HistoricalStateFromSlot(ctx, head.Slot)
```
Comment (terenc3t) [inline beacon-chain/rpc/attester_server.go]: Bug #1: Suppose to use head state here instead of last processed state

File: beacon-chain/rpc/attester_server.go [hunk 2]
```diff
@@ -70,15 +62,12 @@ func (as *AttesterServer) AttestationDataAtSlot(ctx context.Context, req *pb.Att
...
 	if epochStartSlot == head.Slot {
-		hash, err := hashutil.HashBeaconBlock(head)
-		if err != nil {
-			return nil, fmt.Errorf("could not tree hash head block: %v", err)
-		}
-		epochBoundaryRoot = hash[:]
+		epochBoundaryRoot = headRoot[:]
 	} else {
-		epochBoundaryRoot, err = blocks.BlockRoot(beaconState, epochStartSlot)
+		epochBoundaryRoot, err = blocks.BlockRoot(headState, epochStartSlot)
```
Comment (terenc3t) [inline beacon-chain/rpc/attester_server.go]: Bug2: Because of 1, it gives us an incorrect epoch boundary root

File: beacon-chain/rpc/attester_server.go [hunk 3]
```diff
@@ -87,9 +76,9 @@ func (as *AttesterServer) AttestationDataAtSlot(ctx context.Context, req *pb.Att
 	// On the server side, this is fetched by calling get_block_root(state, justified_epoch).
 	// If the last justified boundary slot is the same as state current slot (ex: slot 0),
 	// we set justified block root to an empty root.
-	lastJustifiedSlot := helpers.StartSlot(beaconState.JustifiedEpoch)
+	lastJustifiedSlot := helpers.StartSlot(headState.JustifiedEpoch)
```
Comment (terenc3t) [inline beacon-chain/rpc/attester_server.go]: Bug3: Because of 1, it gives us an incorrect justified slot

File: validator/client/validator_attest.go [hunk 4]
```diff
@@ -68,6 +67,9 @@ func (v *validator) AttestToBlockHead(ctx context.Context, slot uint64) {
 			slot-params.BeaconConfig().GenesisSlot, err)
 		return
 	}
+	// Set the attestation data's slot to head_state.slot where the slot
+	// is the canonical head of the beacon chain.
+	attData.Slot = infoRes.Slot
```
Comment (terenc3t) [inline validator/client/validator_attest.go]: Bug4: attestation slot should be the **head slot**, not the **current slot**

File: proto/beacon/rpc/v1/services.proto [hunk 5]
```diff
@@ -68,6 +68,7 @@ message AttestationDataResponse {
     uint64 justified_epoch = 3;
     bytes justified_block_root_hash32 = 4;
     ethereum.beacon.p2p.v1.Crosslink latest_crosslink = 5;
+    uint64 slot = 6;
```
Comment (prestonvanloon) [inline proto/beacon/rpc/v1/services.proto]: Why is this needed?

Comment (terenc3t) [reply proto/beacon/rpc/v1/services.proto]: Attester needs to know the head slot and it has no one of knowing unless we include head slot in one of the Response

File: shared/p2p/service.go [hunk 6]
```diff
@@ -135,9 +135,9 @@ func (s *Server) Start() {
 		}
 		bcfg := kaddht.DefaultBootstrapConfig
 		bcfg.Period = time.Duration(30 * time.Second)
-		if err := s.dht.BootstrapWithConfig(ctx, bcfg); err != nil {
-			log.Errorf("Failed to bootstrap DHT: %v", err)
-		}
+		//if err := s.dht.BootstrapWithConfig(ctx, bcfg); err != nil {
```
Comment (prestonvanloon) [inline shared/p2p/service.go]: This disables bootstrap discovery. Please remove.

File: beacon-chain/rpc/attester_server.go [hunk 7]
```diff
@@ -46,22 +46,25 @@ func (as *AttesterServer) AttestationDataAtSlot(ctx context.Context, req *pb.Att
...
 	}
-	blockRoot, err := hashutil.HashBeaconBlock(head)
+	headRoot, err := hashutil.HashBeaconBlock(head)
 	if err != nil {
 		return nil, fmt.Errorf("could not tree hash beacon block: %v", err)
 	}
-	beaconState, err := as.beaconDB.State(ctx)
+
+	// Let head state be the state of head block processed through empty slots up to assigned slot.
+	headState, err := as.beaconDB.HistoricalStateFromSlot(ctx, head.Slot)
```
Comment (prestonvanloon) [inline beacon-chain/rpc/attester_server.go]: Please use `as.beaconDB.State(ctx)` and have an assertion that these slots are the same.

File: proto/beacon/rpc/v1/services.proto (hunk 5 above)
Comment (rauljordan) [reply proto/beacon/rpc/v1/services.proto]: can we call this head slot then

File: beacon-chain/rpc/attester_server.go (hunk 7 above)
Comment (terenc3t) [reply beacon-chain/rpc/attester_server.go]: We also don't need the assertion, from the spec:
`Let `head_state` be the state of `head_block` processed through any empty slots up to the assigned slot.`

What we need, what I added is this:
```go
	for headState.Slot < req.Slot {
		headState, err = state.ExecuteStateTransition(
			ctx, headState, nil /* block */, headRoot, state.DefaultConfig(),
		)
		if err != nil {
			return nil, fmt.Errorf("could not execute head transition: %v", err)
		}
	}
```
